geohisto/intercommunalities.py

```diff
@@ -0,0 +1,287 @@
+    """
+    log.debug('Load intercommunalities from %s', filename)
+    validity = datetime(year, 1, 1)
+    intercommunality = Intercommunality()
+    with open(filename) as epci_csv:
+        all_lines = csv.DictReader(epci_csv, delimiter=';', quotechar='"')
+        # No need to sort CSV are already sorted by SIREN
+        # all_lines = sorted(all_lines, key=siren_key)
+
+        for siren, lines in groupby(all_lines, siren_key):
+            for line in lines:
+                if siren != intercommunality.siren:
+                    intercommunality = Intercommunality(
+                        siren=siren,
+                        name=extract_name(line),
+                        acronym=extract_acronym(line),
+                        kind=line['nature'],
+                        taxmodel=line['fiscalite'],
+                        population=line['ptot']
+                    )
+                insee = line['insee'].zfill(5)
+                attach_town(intercommunality, insee, validity, towns)
+            yield intercommunality
+
+
+def load_intercommunalities(towns, directory='sources/epci',
+                            start=INTERCOMMUNALITY_START_YEAR, end=2017):
+    """
+    Load all intercommunalities from directory for the years from start to end
+    """
+    log.info('Processing intercommunalities from %s (%s-%s)',
+             directory, start, end)
+    intercommunalities = Intercommunalities()
+    for year in range(start, end + 1):
+        open_sirens = intercommunalities.open_sirens
+        filename = os.path.join(directory, '{0}.csv'.format(year))
+        for intercommunality in load_intercommunalities_from(filename,
+                                                             year,
+                                                             towns):
+            if intercommunality.siren in open_sirens:
+                # This is either the same or an update
+                intercommunalities.update(intercommunality, year)
+            else:
+                # This is a creation
+                intercommunalities.upsert(intercommunality.create_on(year))
+            open_sirens.discard(intercommunality.siren)
+
+        # All remaining SIRENs are now obsolete intercommunalities
+        for siren in open_sirens:
+            intercommunalities.ends(siren, year - 1, END_REMOVED)
+    return intercommunalities
+
+
+def attach_town(intercommunality, insee, validity, towns):
+    try:
+        intercommunality.towns.add(towns.get_current(insee, validity).id)
+    except:
+        log.error('Failed for %s on %s@%s',
+                  intercommunality.name, insee, validity.isoformat())
+        intercommunality.missing_towns.add((insee, validity))
+
+
+INTERCOMMUNALITY_FIELDS = (
+    'id', 'siren', 'name', 'acronym',
+    'kind', 'taxmodel', 'towns',
+    'start_date', 'end_date', 'end_reason',
+    'successors', 'ancestors',
+    'population',
+)
+
+
+def write_intercommunalities_on(filename, intercommunalities,
```

Comment:
Not sure about it. It was before but it was the only part of intercommunalities process split out in another module so I grouped it here so the whole intercommunalities part is isolated.